packages/permission-controller/src/PermissionController.ts

```diff
@@ -1976,21 +1996,73 @@ export class PermissionController<
    * @param permissionsRequest - The permissions request object.
    * @returns The approved permissions request object.
    */
-  private async requestUserApproval(permissionsRequest: PermissionsRequest) {
+  private async requestUserApproval(
+    permissionsRequest: PermissionsRequest,
+  ): Promise<{
+    approvedRequest: PermissionsRequest;
+    sideEffectsData?: unknown[];
+  }> {
     const { origin, id } = permissionsRequest.metadata;
+
+    const sideEffects = Object.keys(
+      permissionsRequest.permissions,
+    ).reduce<SideEffects>(
+      (sideEffectList, targetName) => {
+        const targetKey = this.getTargetKey(targetName);
+
+        if (targetKey) {
+          const specification = this.getPermissionSpecification(targetKey);
+
+          if (specification.sideEffect) {
+            sideEffectList.permittedHandlers.push(
+              specification.sideEffect.onPermitted,
+            );
+
+            if (specification.sideEffect.onFailure) {
+              if (sideEffectList.failureHandlers) {
+                sideEffectList.failureHandlers.push(
+                  specification.sideEffect.onFailure,
+                );
+              } else {
+                return {
+                  permittedHandlers: sideEffectList.permittedHandlers,
+                  failureHandlers: [specification.sideEffect.onFailure],
+                };
+              }
+            }
+          }
+        }
+        return sideEffectList;
+      },
+      { permittedHandlers: [] },
```

Comment:
Could we simplify by having the empty array for both in the accumulator? That way we can eliminate some ifs.